src/main/scala/org/constellation/primitives/PeerManager.scala

```diff
@@ -134,30 +134,37 @@ object PeerManager extends StrictLogging {
 
   }
 
-  def peerDiscovery(client: APIClient)(implicit dao: DAO): Unit =
-    client
-      .getNonBlocking[Seq[PeerMetadata]]("peers")
-      .onComplete {
-        case Success(pmd) =>
-          pmd.foreach { md =>
-            if (dao.id != md.id && validPeerAddition(HostPort(md.host, md.httpPort), dao.peerInfo.unsafeRunSync())) {
-              val client =
-                APIClient(md.host, md.httpPort)(dao.apiClientExecutionContext, dao)
-              client
-                .getNonBlocking[PeerRegistrationRequest]("registration/request")
-                .onComplete {
-                  case Success(registrationRequest) =>
-                    dao.peerManager ! PendingRegistration(md.host, registrationRequest)
-                    client.post("register", dao.peerRegistrationRequest)
-                  case Failure(e) =>
-                    dao.metrics.incrementMetric("peerGetRegistrationRequestFailed")
-                }(dao.apiClientExecutionContext)
-            }
-          }
-        case Failure(e) =>
-          dao.metrics.incrementMetric("peerDiscoveryQueryFailed")
+  def peerDiscovery(client: APIClient)(implicit dao: DAO): Unit = {
 
-      }(dao.apiClientExecutionContext)
+    implicit val ec = dao.apiClientExecutionContext
+
+    val discover: IO[List[String]] = for {
+      peersMetadata <- client.getNonBlockingIO[Seq[PeerMetadata]]("peers").recoverWith {
+        case err =>
+          dao.metrics.incrementMetricAsync[IO]("peerDiscoveryQueryFailed") *> IO.raiseError(err)
+      }
+      peers <- dao.peerInfo
+      filteredPeers <- IO(
+        peersMetadata.filter(p => p.id != dao.id && validPeerAddition(HostPort(p.host, p.httpPort), peers))
+      )
+      register <- filteredPeers.toList
+        .traverse(
+          md =>
+            APIClient(md.host, md.httpPort)
+              .getNonBlockingIO[PeerRegistrationRequest]("registration/request")
+              .map(req => (md, req))
+        )
+      _ <- IO(register.foreach(r => dao.peerManager ! PendingRegistration(r._1.host, r._2)))
```

Comment:
why not `register.traverse(r => IO(dao.peerManager ! PendingRegistration(r._1.host, r._2)))`?